File: apps/burner-mac/Sources/FlagshipBurnerCore/IsoManifestClient.swift
import Foundation

/// Talks to the server's ISO-manifest endpoint to decide which base ISO the
/// burner should hold. The burner is a DUMB EXECUTOR: it reports the version +
/// sha256 of whatever it has cached (or `nil` if it has nothing) and the server
/// replies with either an order to download a specific URL/sha/version or `null`
/// meaning "keep what you've got". The client never compares shas to decide;
/// it just obeys the order and then verifies the bytes it downloads.
///
/// LOCKED WIRE CONTRACT:
///   POST https://flagshipserver.com/api/iso-manifest
///   Request:  { "platform":"mac", "burnerVersion":"<s>",
///               "current": { "version":"<s>", "sha256":"<hex64>" } | null }
///   Response: { "download": { "url","sha256","version","sizeBytes","attestation" } }
///        or:  { "download": null }

/// What the burner currently holds in its cache, reported verbatim to the server.
public struct IsoManifestCurrent: Codable, Equatable, Sendable {
    public let version: String
    public let sha256: String

    public init(version: String, sha256: String) {
        self.version = version
        self.sha256 = sha256
    }
}

/// The request body POSTed to /api/iso-manifest.
public struct IsoManifestRequest: Codable, Equatable, Sendable {
    public let platform: String
    public let burnerVersion: String
    public let current: IsoManifestCurrent?

    public init(platform: String = "mac", burnerVersion: String, current: IsoManifestCurrent?) {
        self.platform = platform
        self.burnerVersion = burnerVersion
        self.current = current
    }
}

/// The server's "download this exact ISO" order.
public struct IsoManifestDownload: Codable, Equatable, Sendable {
    public let url: String
    public let sha256: String
    public let version: String
    public let sizeBytes: Int
    public let attestation: String

    public init(url: String, sha256: String, version: String, sizeBytes: Int, attestation: String) {
        self.url = url
        self.sha256 = sha256
        self.version = version
        self.sizeBytes = sizeBytes
        self.attestation = attestation
    }
}

/// The response body. Exactly one shape: `download` is either an order or null.
public struct IsoManifestResponse: Codable, Equatable, Sendable {
    public let download: IsoManifestDownload?

    public init(download: IsoManifestDownload?) {
        self.download = download
    }
}

public enum IsoManifestError: LocalizedError {
    case offline(String)
    case httpStatus(Int)
    case decode(String)

    public var errorDescription: String? {
        switch self {
        case .offline(let why):
            return "Couldn't reach the base-image service — check your internet connection. (\(why))"
        case .httpStatus(let code):
            return "The base-image service returned HTTP \(code)."
        case .decode(let why):
            return "The base-image service sent a response we couldn't read. (\(why))"
        }
    }
}

public struct IsoManifestClient: Sendable {
    public static let endpoint = URL(string: "https://flagshipserver.com/api/iso-manifest")!

    private let endpoint: URL
    private let session: URLSession

    public init(endpoint: URL = IsoManifestClient.endpoint, session: URLSession = .shared) {
        self.endpoint = endpoint
        self.session = session
    }

    /// POST the manifest request and decode the reply. Throws `IsoManifestError`
    /// on transport, HTTP, or decode failure.
    public func fetch(_ request: IsoManifestRequest) async throws -> IsoManifestResponse {
        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 30
        do {
            req.httpBody = try JSONEncoder().encode(request)
        } catch {
            throw IsoManifestError.decode("could not encode request: \(error.localizedDescription)")
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch let e as URLError {
            throw IsoManifestError.offline(e.localizedDescription)
        }
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw IsoManifestError.httpStatus(http.statusCode)
        }
        do {
            return try JSONDecoder().decode(IsoManifestResponse.self, from: data)
        } catch {
            throw IsoManifestError.decode(error.localizedDescription)
        }
    }
}

File: apps/burner-linux/iso_manifest_client.py
"""Server-manifest client for the Simple-mode Debian base ISO.

The burner is a DUMB EXECUTOR. It POSTs its current cached-base state to the
server's /api/iso-manifest endpoint and obeys the response verbatim:

  - response has a non-null `download` → fetch + verify + cache those bytes.
  - response `download` is null → keep whatever is cached.

The burner NEVER decides by sha client-side; it only *reports* `current` and
*verifies* the bytes it downloads against the sha the server hands back.

LOCKED WIRE CONTRACT
--------------------
POST https://flagshipserver.com/api/iso-manifest

Request JSON:
  { "platform": "linux",
    "burnerVersion": "<string>",
    "current": { "version": "<string>", "sha256": "<hex64>" } | null,
    "arch": "amd64" | "arm64" }          # OPTIONAL; absent = "amd64"

Burning always requests amd64 (real boxes are x86; the arch key stays ABSENT
so the burn-path request is byte-identical to the pre-arch wire format). The
host-on-this-PC path requests the HOST arch — an arm64 guest is the only kind
an arm64 Chromebook/SBC can host. `download: null` for a requested arch means
"keep what you've got" — UNLESS nothing is cached for that arch, in which case
the CACHE layer surfaces "no base for this arch" (see iso_base_cache).

Response JSON, exactly one of:
  { "download": { "url": "<https>", "sha256": "<hex64>",
                  "version": "<string>", "sizeBytes": <int>,
                  "attestation": "<https>" } }
  { "download": null }

Stdlib only (urllib + json) so it imports without GTK or third-party packages
and works inside the AppImage/Flatpak sandbox.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

PLATFORM = "linux"
MANIFEST_URL = "https://flagshipserver.com/api/iso-manifest"


class ManifestError(Exception):
    """Base for manifest-fetch failures — carries a user-facing message."""


class ManifestOfflineError(ManifestError):
    def __init__(self, why: str) -> None:
        super().__init__(
            f"Couldn't reach the base-image manifest — check your internet "
            f"connection. ({why})"
        )


class ManifestHTTPError(ManifestError):
    def __init__(self, code: int) -> None:
        self.code = code
        super().__init__(f"Base-image manifest request failed (HTTP {code}).")


class ManifestParseError(ManifestError):
    def __init__(self, why: str) -> None:
        super().__init__(f"Base-image manifest response was malformed. ({why})")


@dataclass(frozen=True)
class CurrentBase:
    """What the burner currently has cached. Reported as `current`."""
    version: str
    sha256: str

    def to_wire(self) -> dict:
        return {"version": self.version, "sha256": self.sha256}


@dataclass(frozen=True)
class DownloadDescriptor:
    """The server's instruction to fetch a new base ISO."""
    url: str
    sha256: str
    version: str
    size_bytes: int
    attestation: str

    @classmethod
    def from_wire(cls, d: dict) -> "DownloadDescriptor":
        try:
            parsed = cls(
                url=str(d["url"]),
                sha256=str(d["sha256"]).lower(),
                version=str(d["version"]),
                size_bytes=int(d["sizeBytes"]),
                attestation=str(d["attestation"]),
            )
        except (KeyError, TypeError, ValueError) as e:
            raise ManifestParseError(f"download missing/invalid field: {e}") from e
        # Match the Windows/Mac clients: refuse a non-https byte source and a
        # sha that can't possibly verify (the download would fail anyway, but
        # fail HERE with a message that blames the manifest, not the network).
        if not parsed.url.startswith("https://"):
            raise ManifestParseError("download URL is not https")
        if len(parsed.sha256) != 64 or any(
            c not in "0123456789abcdef" for c in parsed.sha256
        ):
            raise ManifestParseError("sha256 is not a 64-char hex digest")
        return parsed


@dataclass(frozen=True)
class ManifestResult:
    """Parsed /api/iso-manifest response. `download` is None when the server
    tells the burner to keep its cache."""
    download: Optional[DownloadDescriptor]


def fetch_manifest(
    burner_version: str,
    current: Optional[CurrentBase],
    url: str = MANIFEST_URL,
    *,
    arch: Optional[str] = None,
    opener=None,
) -> ManifestResult:
    """POST the manifest request and parse the response.

    `opener` is an injectable seam for tests: a callable taking a
    urllib.request.Request and returning a file-like response (mirrors
    urlopen). Defaults to the real urlopen.

    Raises ManifestError subclasses on network / HTTP / parse failure.
    """
    do_open = opener or (lambda req: urlopen(req))  # noqa: S310 - fixed https URL constant

    body = {
        "platform": PLATFORM,
        "burnerVersion": burner_version,
        "current": current.to_wire() if current is not None else None,
    }
    # amd64 is encoded as an ABSENT key so the burn-path request stays
    # byte-identical to the pre-arch wire format (matches the Mac client).
    if arch is not None and arch != "amd64":
        body["arch"] = arch
    payload = json.dumps(body).encode("utf-8")
    req = Request(
        url,
        data=payload,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "flagship-burner-linux",
        },
    )

    try:
        resp = do_open(req)
    except HTTPError as e:
        raise ManifestHTTPError(e.code) from e
    except (URLError, OSError) as e:
        raise ManifestOfflineError(str(getattr(e, "reason", e))) from e

    try:
        with resp:
            status = getattr(resp, "status", 200) or 200
            if not (200 <= status <= 299):
                raise ManifestHTTPError(status)
            raw = resp.read()
    except (URLError, OSError) as e:
        raise ManifestOfflineError(str(getattr(e, "reason", e))) from e

    try:
        parsed = json.loads(raw)
    except (ValueError, json.JSONDecodeError) as e:
        raise ManifestParseError(str(e)) from e

    if not isinstance(parsed, dict) or "download" not in parsed:
        raise ManifestParseError("response is not an object with a `download` key")

    dl = parsed["download"]
    if dl is None:
        return ManifestResult(download=None)
    if not isinstance(dl, dict):
        raise ManifestParseError("`download` must be an object or null")
    return ManifestResult(download=DownloadDescriptor.from_wire(dl))
